// dynamodbHelper/index.js
const AWS = require('aws-sdk');

/**
 * Get DynamoDB DocumentClient instance
 * @returns {AWS.DynamoDB.DocumentClient}
 */
function getDocClient() {
    return new AWS.DynamoDB.DocumentClient();
}

/**
 * Query items from DynamoDB table
 * @param {string} tableName - Table name
 * @param {string} keyConditionExpression - Key condition expression
 * @param {object} expressionAttributeValues - Expression attribute values
 * @param {string} indexName - Optional index name
 * @returns {Promise<Array>} Array of items
 */
async function queryItems(tableName, keyConditionExpression, expressionAttributeValues, indexName = null) {
    const docClient = getDocClient();
    const params = {
        TableName: tableName,
        KeyConditionExpression: keyConditionExpression,
        ExpressionAttributeValues: expressionAttributeValues
    };
    if (indexName) {
        params.IndexName = indexName;
    }
    const data = await docClient.query(params).promise();
    return data.Items;
}

/**
 * Get item by primary key
 * @param {string} tableName - Table name
 * @param {object} key - Primary key object
 * @returns {Promise<object|null>} Item or null if not found
 */
async function getItem(tableName, key) {
    const docClient = getDocClient();
    const params = {
        TableName: tableName,
        Key: key
    };
    const data = await docClient.get(params).promise();
    return data.Item || null;
}

/**
 * Put item into DynamoDB table
 * @param {string} tableName - Table name
 * @param {object} item - Item to put
 * @returns {Promise<void>}
 */
async function putItem(tableName, item) {
    const docClient = getDocClient();
    const params = {
        TableName: tableName,
        Item: item
    };
    await docClient.put(params).promise();
}

/**
 * Update item in DynamoDB table
 * @param {string} tableName - Table name
 * @param {object} key - Primary key object
 * @param {string} updateExpression - Update expression
 * @param {object} expressionAttributeValues - Expression attribute values
 * @param {string} returnValues - Return values (default 'ALL_NEW')
 * @returns {Promise<object>} Updated attributes
 */
async function updateItem(tableName, key, updateExpression, expressionAttributeValues, returnValues = 'ALL_NEW') {
    const docClient = getDocClient();
    const params = {
        TableName: tableName,
        Key: key,
        UpdateExpression: updateExpression,
        ExpressionAttributeValues: expressionAttributeValues,
        ReturnValues: returnValues
    };
    const data = await docClient.update(params).promise();
    return data.Attributes;
}

/**
 * Delete item from DynamoDB table
 * @param {string} tableName - Table name
 * @param {object} key - Primary key object
 * @returns {Promise<void>}
 */
async function deleteItem(tableName, key) {
    const docClient = getDocClient();
    const params = {
        TableName: tableName,
        Key: key
    };
    await docClient.delete(params).promise();
}

/**
 * Scan all items from DynamoDB table
 * @param {string} tableName - Table name
 * @returns {Promise<Array>} Array of items
 */
async function scanItems(tableName) {
    const docClient = getDocClient();
    const params = {
        TableName: tableName
    };
    const data = await docClient.scan(params).promise();
    return data.Items;
}

module.exports = {
    getDocClient,
    queryItems,
    getItem,
    putItem,
    updateItem,
    deleteItem,
    scanItems
};
