// internalClient/index.js
const AWS = require('aws-sdk');

/**
 * Invoke another Lambda function using AWS SDK
 * @param {string} functionName - Target Lambda function name (from environment variable)
 * @param {object} payload - Payload to send (will be JSON serialized)
 * @param {string} invocationType - 'RequestResponse' (sync) or 'Event' (async). Default: 'RequestResponse'
 * @returns {Promise<object>} Parsed response payload for RequestResponse, null for Event
 */
async function invokeLambda(functionName, payload, invocationType = 'RequestResponse') {
    const lambda = new AWS.Lambda();
    const params = {
        FunctionName: functionName,
        InvocationType: invocationType,
        Payload: JSON.stringify(payload)
    };
    
    try {
        const response = await lambda.invoke(params).promise();
        if (invocationType === 'RequestResponse') {
            if (response.Payload) {
                return JSON.parse(response.Payload.toString());
            }
            return null;
        }
        return null;
    } catch (error) {
        console.error(`Error invoking Lambda ${functionName}:`, error);
        throw error;
    }
}

module.exports = {
    invokeLambda
};
