// responseHelper/index.js
/**
 * Standard API response helper
 */

function success(body, statusCode = 200) {
    return {
        statusCode,
        headers: {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Credentials': true
        },
        body: JSON.stringify(body)
    };
}

function error(message, statusCode = 500) {
    return {
        statusCode,
        headers: {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Credentials': true
        },
        body: JSON.stringify({ error: message })
    };
}

function notFound(message = 'Resource not found') {
    return error(message, 404);
}

function badRequest(message = 'Bad request') {
    return error(message, 400);
}

function unauthorized(message = 'Unauthorized') {
    return error(message, 401);
}

module.exports = {
    success,
    error,
    notFound,
    badRequest,
    unauthorized
};
